### PR TITLE
Support fork points with listed and unlisted commits

### LINKED ISSUE #449: Add support for overriding fork points - listed and unlisted commits
Issue #307 covers the following cases:
- overriding to parent
- overriding to inferred

***
## The scope of this task

Enhance the `Override Fork Point` Dialog with an option (aka radioButton) containing a combobox with a list of commits.
We should allow overriding fork point to one of the branch commits (listed) and of the older ones (not listed).

For UI see: https://jetbrains.org/intellij/sdk/docs/user_interface_components/kotlin_ui_dsl.html#combo-boxes

### PR DESCRIPTION
**Pull request by vzmerr**
It resolves #449 

Should we add a custom renderer to render each commit with its hash and message?

### CODE REVIEW

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 1]
```diff
@@ -92,5 +95,17 @@ class OverrideForkPointDialog(
         MutableProperty(::myOverrideOption) { myOverrideOption = it },
         OverrideOption::class.java
       )
+
+    row("The fork point commit:") {
+      comboBox(
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:100]: Hmmm for the sake of consistency... would it be possible to rather have a single `radioButton` for each commit? with parent comment and inferred commit highlighted as suggested options

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:106]: Do we still need the existing descriptions in front of the radioButtons?

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:85]: 🤔  could you attach the screenshot of how the entire dialog looks for the current PR state?

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:85]: I'm trying to get it compile, first. It seems that some null value in the combobox is getting passed to the  overrided method of the custom Renderer, which leads to a null pointer exception. I am trying to debug that first.

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:85]: currently, at runtime, some null value is getting passed from the combobox to the overrided method of the cell renderer, leading to a null pointer exception.

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:74]: image

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: image
latest

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: Okay... @mkondratek as you're a local UXer, could you suggest where could go from there? I've got some ideas but don't want to suggest anything yet

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: Hey @mkondratek 👋

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: whoops, Idk why but I did not receive email notifications from this one. maybe I must be added as a reviewer? dunno. feel free to ping me directly sooner if I am not responding 😆

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: Hmmm, I think that the parent branch and the inferred commit are usually the right ones and we should not hide them among the other commits. I would rather keep them.

My idea is to have 3 radio buttons:
 [ ] parent branch 
 [ ] inferred commit
 [ ] manually pointed commit <combo box with all commits>


wdyt?

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: ☝🏻 It was indeed the original Ina's idea... although it also looked confusing to me (as inferred commit is then displayed twice: first highlighted, and then listed among other non-highlighted commits).

I'd rather see **a single list**, but:
1. scrollable
2. with some rich(er) formatting (not just combobox with plain unformatted text)
3. with parent branch and inferred commit clearly highlighted

WDYT?

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: Sometimes also the selected commit in the combobox is invalid as a fork point.

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: In the latest commit of this branch, the inferred commit and the parent one are at the top of the list of the combobox. and labeled as parent and inferred. WDYT? what can we do about the richer format or can we put them separately?

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: @PawelLipski I am not convinced to this idea. I would rather list some commits twice than abandon the clear indication of what are the most possible commits to pick

and I cannot imagine that list containing all that information within. I mean - ofc we can bold smth, add `[tag-telling-what-I-am]` etc, but I doubt it will remain looking clearly and nice

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: > Sometimes also the selected commit in the combobox is invalid as a fork point. We should be filtering out such commits, shouldn't we?

Hmmm what specifically do you mean by _invalid as fork point_? 🤔

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: > I am not convinced to this idea. I would rather list some commits twice than abandon the clear indication of what are the most possible commits to pick

Okay, you've got me convinced/outvoted 😅  Let's go 3 radio buttons then, and some list (not necessarily combobox, better sth that allows for rich formatting) for the _manually pointed commit_ option

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: > not necessarily combobox, better sth that allows for rich formatting

@mkondratek any recommendations?

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: @PawelLipski sorry, I wanted to reply to your comment, and by mistake I ended up editing it somehow 😬

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: > Hmmm what specifically do you mean by invalid as fork point? 🤔

I would repro the error in the plugin and let you know

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: > > not necessarily combobox, better sth that allows for rich formatting
> 
> @mkondratek any recommendations?

@vzmerr idk, while looking for components/ui ideas I often depend on what I remember from the IDE, find it out in the intellij code and apply similar things in our code. you may want to take a look into https://plugins.jetbrains.com/docs/intellij/kotlin-ui-dsl-version-2.html too. for me combobox sounds most reasonable.

iiuc @PawelLipski would be happier with a list (seems to be anti-patternish to me, but anyway). this idea basicly expands to radiobutton set within radiobutton set:

(here checkboxes but single option required ofc)
- [ ] fork point
- [ ] parent
- [ ] manually picked commit
  - [ ] abc2137 Do a lot of things
  - [ ] 2a1b3c7 Do a lot of things - 1st round of fixes
  - [ ] 020897 Do a lot of things - final commit

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: > list (seems to be anti-patternish to me, but anyway

Could you elaborate? 🤔

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: Where have you seen a list used like that? can you bring some examples of your idea applied? from intellij? of from some website (https://www.intercity.pl/pl/ does not count 😆 )?

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: 😅 okay maybe actually a scrollable list with radio boxes isn't the best idea, let's stick to combobox indeed.

I'd suggest another thing however (can be done within a separate issue): to display, within the Override Fork Point dialog, **the list of commits that will be considered unique to the branch, assuming the select fork point**... WDYT?

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: image
The latest look. WDYT?

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: Looks okay to me... probably the **subscript** under _Manually picked commit_ is redundant tho

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 2]
```diff
@@ -87,6 +100,99 @@ class OverrideForkPointDialog(
...
+            private val defaultBackground = UIManager.get("List.background") as Color
+            override fun getListCellRendererComponent(
+              list: JList<*>?,
+              value: Any?,
+              index: Int,
+              isSelected: Boolean,
+              cellHasFocus: Boolean
+            ): Component {
+              val commit: ICommitOfManagedBranch? = value as ICommitOfManagedBranch?
+              componentOrientation = list!!.componentOrientation
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:130]: Hmm what's the guarantee that `list` is non-null? 🤔

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 3]
```diff
@@ -87,6 +100,99 @@ class OverrideForkPointDialog(
...
+                background = list.background
+                foreground = list.foreground
+              }
+
+              icon = null
+
+              isEnabled = list.isEnabled
+
+              text = if (commit != null) {
+                var prefix =
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:158]: Hmmm currently the combo box list looks like that:

image

for list of commits like

image

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:158]: I'd rather avoid including _Inferred commit_ at the beginning of the list by default (?)

Also, let's order the commits in the same order as they appear in the branch status: from oldest ("closest" to the parent branch) to the newest ("closest" to the child branch).

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 4]
```diff
@@ -87,6 +100,99 @@ class OverrideForkPointDialog(
             thisRowComment
           )
       }
+
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:103]: The more I look at that, the more I think that a list of commits **unique** to the branch (i.e. what's gonna be displayed in the status for the branch) would be useful, right above or below the radio buttons... WDYT folks?

Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:103]: @PawelLipski 👌

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt (hunk 2 above)
Comment (vzmerr) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:130]: we're the one constructing it initially with the list of commits.

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt (hunk 3 above)
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:158]: yup, lets keep the order

**Review by mkondratek — CHANGES_REQUESTED** [review]
image

1. I am against including `infered commit (branch)` and `parent branch (branch)` texts - it is redundant - these info is already given by the above radio options.  

2. If we are going to keep them then a space between the text and commit hash is missing - pls add it

3. There is inaccurate comment below the new radio option - pls fix it (probably remove); example below
image

**Comment by mkondratek** [comment]
In spite of the above issues, this feature looks nice 🙂 I think we could have it soon

**Comment by PawelLipski** [comment]
@zmerr if you could wrap up this PR before proceeding with traverse (as this PR requires like an order of magnitude less work to complete 😅)

**Comment by mkondratek** [comment]
@vzmerr I am gonna take this PR from you :)

**Comment by PawelLipski** [comment]
☝🏻 pls use @zmerr, @vzmerr is no longer active

If you Ina have some unpushed work her since y'day, pls push

**Review by mkondratek — APPROVED** [review]
image

**Comment by mkondratek** [comment]
I think it looks pretty good

image

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseOverrideForkPointAction.java [hunk 5]
```diff
@@ -122,7 +122,7 @@ private void setOverrideForkPointConfigValues(
     }
 
     try {
-      GitConfigUtil.setValue(project, root, whileDescendantOfKey, ancestorCommit.getHash());
+      GitConfigUtil.setValue(project, root, whileDescendantOfKey, forkPoint.getHash());
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/base/BaseOverrideForkPointAction.java:125]: This is not strictly related to the change at hand, it's a step towards getting rid of `whileDescendantOf` completely (https://github.com/VirtusLab/git-machete-intellij-plugin/issues/1030)

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 6]
```diff
@@ -3,15 +3,24 @@ package com.virtuslab.gitmachete.frontend.actions.dialogs
 import com.intellij.openapi.project.Project
```
[the remark this answers is not in the document]
Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:3]: Let's assume that v3.1.1 will be released separately before v3.2.0

[the remark this answers is not in the document]
Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:3]: Plus bump the version to 3.2.0

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 7]
```diff
@@ -87,6 +100,77 @@ class OverrideForkPointDialog(
...
+            getString(
+              "action.GitMachete.BaseOverrideForkPointAction.dialog.override-fork-point.radio-button.custom"
+            ),
+            radioButtonComment
+          ),
+          OverrideOption.CUSTOM
+        )
+
+        comboBox<ICommitOfManagedBranch?>(
+          (listOf(branch.forkPoint, parentBranch.pointedCommit) + branch.commits.toMutableList()).filterNotNull(),
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:116]: Hmmm there's a problem with taking `branch.commits` as a range of commits here, as `commits` already take existing fork point override into account (if one exists 😅)

As a quick workaround, I'd:
* add `commitsUntilParent` alongside `commits` to `NonRootManagedBranchSnapshot`, computed as `gitCoreRepository.deriveCommitRange(corePointedCommit, parentCoreLocalBranch.getPointedCommit())` regardless of `syncToParentStatus`
* rename `commits` to `uniqueCommits`
* use `branch.commitsUntilParent` instead of `branch.(unique)commits`

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:116]: (preferably on a PR under the current one)

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: done here: https://github.com/VirtusLab/git-machete-intellij-plugin/pull/1273

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 8]
```diff
@@ -87,6 +91,42 @@ class OverrideForkPointDialog(
...
+          object : DefaultListCellRenderer() {
+            override fun getListCellRendererComponent(
+              list: JList<*>?,
+              value: Any?,
+              index: Int,
+              isSelected: Boolean,
+              cellHasFocus: Boolean
+            ): Component {
+              val commit: ICommitOfManagedBranch = value as ICommitOfManagedBranch
+              text = "<html><tt>[${commit.shortHash}]</tt> ${commit.shortMessage}</html>"
```
Comment (mkondratek) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:122]: image

monospace font for commit hashes - same len of them, better look imho

File: CHANGE-NOTES.md [hunk 9]
```diff
@@ -3,6 +3,9 @@
 ## Unreleased
 - Removed automatic rediscovery in case of empty machete file
```
Comment (PawelLipski) [inline CHANGE-NOTES.md:4]: Move this to v3.1.1 below

Comment (PawelLipski) [reply CHANGE-NOTES.md:4]: Aaand lemme open a task to get rid of this section ✂️

Comment (PawelLipski) [reply CHANGE-NOTES.md:4]: #1274

Comment (mkondratek) [reply CHANGE-NOTES.md:4]: ok

Comment (PawelLipski) [reply CHANGE-NOTES.md]: _słuchaj... okej_

**Comment by PawelLipski** [comment]
Wow _it's so f**king ~big~ wide_:

image

Use commit subject, not full message

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt (hunk 8 above)
Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:122]: Lemme comment here so that it's easier to track:

Wow _it's so f**king ~big~ wide_:

image

Use commit subject, not full message

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:122]: Although huh, it's `shortMessage`... hm WTF

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:122]: Awww I see, check `git show 2a09`:

image

and now check the javadoc for `org.eclipse.jgit.revwalk.RevCommit#getShortMessage`:

image

It's a general convention that the **second line** of commit message, if exists, is always empty (as a separating between subject and extended description).

However, judging by the following line in the implementation of `getShortMessage`:

```java
                if (hasLF(raw, msgB, msgE)) {
			str = StringUtils.replaceLineBreaksWithSpace(str);
		}
```

they're indeed aware that the subject (part until first double LF) itself might be multiline.

File: frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties [hunk 10]
```diff
@@ -64,6 +64,7 @@ action.GitMachete.BaseOverrideForkPointAction.dialog.override-fork-point.label.H
 action.GitMachete.BaseOverrideForkPointAction.dialog.override-fork-point.title=Override Fork Point
 action.GitMachete.BaseOverrideForkPointAction.dialog.override-fork-point.radio-button.parent=Parent branch ({0})
 action.GitMachete.BaseOverrideForkPointAction.dialog.override-fork-point.radio-button.inferred=Inferred commit ({0})
+action.GitMachete.BaseOverrideForkPointAction.dialog.override-fork-point.radio-button.custom=Manually picked commit:
```
Comment (PawelLipski) [inline frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties:67]: Also, I'd automatically switch radio to `custom` if a choice is made within the combobox... as for now, one can end up making a choice in combobox, but the radio button remains unchanged

Comment (PawelLipski) [reply frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties:67]: I can think of a formerly existing place in IntelliJ which such the same UX glitch... coincidentally, also related to _overriding_ things 😅 

image

Fixed in 2022.1 or 2022.2, but in the past you could select association in combobox, and nothing happened ☹️

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt (hunk 8 above)
Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:122]: #1277

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 11]
```diff
@@ -50,28 +59,23 @@ class OverrideForkPointDialog(
         )
       )
     }
+
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:62]: Re-posting so that it's not lost:

> I'd suggest another thing however (can be done within a separate issue): to display, within the Override Fork Point dialog, the list of commits that will be considered unique to the branch, assuming the selected fork point... WDYT?

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 12]
```diff
@@ -87,6 +91,42 @@ class OverrideForkPointDialog(
...
+            ),
+            radioButtonComment
+          ),
+          OverrideOption.CUSTOM
+        )
+
+        val mutableList = branch.commitsUntilParent.toMutableList()
+        mutableList.remove(branch.forkPoint)
+        mutableList.remove(parent.pointedCommit)
+        val items = (listOf(branch.forkPoint, parent.pointedCommit) + mutableList.reversed()).filterNotNull()
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:109]: Whoops I'd rather keep the order of commits as-is, without putting inferred fork point & parent as first items... that's very confusing (I just tried to match the order from combobox to git history and could get it ;p)

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 13]
```diff
@@ -87,6 +91,42 @@ class OverrideForkPointDialog(
...
+          items,
+          object : DefaultListCellRenderer() {
+            override fun getListCellRendererComponent(
+              list: JList<*>?,
+              value: Any?,
+              index: Int,
+              isSelected: Boolean,
+              cellHasFocus: Boolean
+            ): Component {
+              val commit: ICommitOfManagedBranch = value as ICommitOfManagedBranch
```
[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:121]: image

[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:121]: whoops xd

[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:121]: ok, will be fixed and look like that:

image

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt (hunk 11 above)
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:62]: IIUC that would be a new look - not something we want to simply add to the present dialog, right?

anyway, instead of re-posting new ideas (or more critical changes) in resolvable and collapsable comments within a PR it is better to discuss them within the related issue or to open a new one, so that it's not lost 🙂

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:62]: https://github.com/VirtusLab/git-machete-intellij-plugin/issues/1279

File: frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties (hunk 10 above)
Comment (mkondratek) [reply frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties:67]: done! 

image

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt (hunk 12 above)
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:107]: ok so to my understanding the most reasonable order is determined by these rules:
1. unique commits until parent in the same order as in git machete tree
2. fork point 
  - if fork point **is** present among the unique commits until parent then it remains in its place
  - if fork point **is not** present among the unique commits until parent then it is added BEFORE them
3. parent - the same rules as fork point (but applied later) 

@PawelLipski, is it correct?

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:107]: Okay, let's go with that 👍🏻

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 14]
```diff
@@ -79,14 +81,50 @@ class OverrideForkPointDialog(
...
               "action.GitMachete.BaseOverrideForkPointAction.dialog.override-fork-point.radio-button.inferred"
             ),
-            radioButtonComment
+            branch.forkPoint?.shortHash ?: "cannot resolve commit hash"
           ),
           OverrideOption.INFERRED
         )
           .comment(
-            thisRowComment
+            escapeHtml4(branch.forkPoint?.shortMessage) ?: "cannot resolve commit message"
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:89]: For `null` input param, what does `escapeHtml4` return? also `null` (no NPE)?

Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:89]: null - I ve check that before. I should have mention it here in the comment that I am aware of what I am doing

Comment (PawelLipski) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: Okay if so then pls mark both param and return type of `escapeHtml4` as `@PolyNull` in stubs, see https://checkerframework.org/manual/#handling-warnings-step2

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 15]
```diff
@@ -79,13 +81,45 @@ class OverrideForkPointDialog(
...
+          format(
+            getString(
+              "action.GitMachete.BaseOverrideForkPointAction.dialog.override-fork-point.radio-button.custom"
+            ),
+            branch.forkPoint?.shortHash ?: "cannot resolve commit hash"
+          ),
+          OverrideOption.CUSTOM
+        )
+
+        val list = branch.commitsUntilParent.reverse() + listOf(branch.forkPoint, parent.pointedCommit)
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:102]: As discussed... shouldn't the parent be at top (==beginning of the list)? to match the top-to-down==oldest-to-latest ordering from status 🤔

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt [hunk 16]
```diff
@@ -2,25 +2,34 @@ package com.virtuslab.gitmachete.frontend.actions.dialogs
...
-  private val parentBranch: IManagedBranchSnapshot,
   private val branch: INonRootManagedBranchSnapshot
 ) : DialogWrapper(project, /* canBeParent */ true) {
 
   private var myOverrideOption = OverrideOption.PARENT
+  private var customCommit: ICommitOfManagedBranch? = branch.forkPoint
+
+  private val parent = branch.parent
+
+  private lateinit var rb: Cell<JBRadioButton>
```
Comment (PawelLipski) [inline frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt:32]: Can it be just a local val? 🤔

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt (hunk 15 above)
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: good point. check it now pls

File: frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt (hunk 14 above)
Comment (mkondratek) [reply frontend/actions/src/main/kotlin/com/virtuslab/gitmachete/frontend/actions/dialogs/OverrideForkPointDialog.kt]: done

File: frontend/ui/impl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/GitMacheteStorage.java [hunk 17]
```diff
@@ -0,0 +1,6 @@
+package com.virtuslab.gitmachete.frontend.ui.impl.table;
+
+import com.intellij.openapi.components.Storage;
+
+@Storage("git-machete.xml")
```
Comment (PawelLipski) [inline frontend/ui/impl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/GitMacheteStorage.java:5]: Wow that's it? 😯

Comment (PawelLipski) [reply frontend/ui/impl/src/main/java/com/virtuslab/gitmachete/frontend/ui/impl/table/GitMacheteStorage.java:5]: Not used on this PR so far, right?

**Review by PawelLipski — APPROVED** [review]
🍴

**Review by PawelLipski — APPROVED** [review]

**Comment by mkondratek** [comment]
@zmerr, thank you for starting and doing most of the work with this one!

**Comment by zmerr** [comment]
> @zmerr, thank you for starting and doing most of the work with this one!

Thanks Mikolaj, but I guess you were the one who ended up doing most of the hard work 😊🎊congrats🎉
